streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java

```diff
@@ -2484,6 +2511,48 @@ public void shouldUpdateOffsetIfCorruptedRecordFollowsValid() {
         );
     }
 
+    @Test
+    public void shouldCheckpointAfterRestorationWhenAtLeastOnceEnabled() {
+        final ProcessorStateManager processorStateManager = mockStateManager();
+        recordCollector = mock(RecordCollectorImpl.class);
+        doReturn(singletonMap(partition1, 1L)).when(recordCollector).offsets();
+
+        task = createStatefulTask(createConfig(AT_LEAST_ONCE, "100"), true, processorStateManager);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        verify(processorStateManager, times(1)).checkpoint();
+        verify(recordCollector, times(1)).offsets();
+    }
+
+    @Test
+    public void shouldNotCheckpointAfterRestorationWhenExactlyOnceEnabled() {
+        final ProcessorStateManager processorStateManager = mockStateManager();
+        recordCollector = mock(RecordCollectorImpl.class);
+
+        task = createStatefulTask(createConfig(EXACTLY_ONCE_V2, "100"), true, processorStateManager);
+        task.initializeIfNeeded();
+        task.completeRestoration(noOpResetter -> { });
+        verify(processorStateManager, never()).checkpoint();
+        verify(processorStateManager, never()).changelogOffsets();
+        verify(recordCollector, never()).offsets();
+    }
+
+    private ProcessorStateManager mockStateManager() {
+        final ProcessorStateManager manager = spy(new ProcessorStateManager(
```

Comment:
I'm doing a partial mock here, to make sure the correct methods are being invoked.

Reply:
I am afraid I cannot follow here. With Mockito you can define strict stubbings. That means Mockito will throw an exception if you have defined stubs that are not used. To use strict mocks, we usually specify 
```java
@MockitoSettings(strictness = Strictness.STRICT_STUBS)
```
on the test class level. Here, this does not work because I think (I am not sure) we use
```java
@RunWith(EasyMockRunner.class)
```    
What works is using
```java
    @Before
    public void setup() {
        mockito = Mockito.mockitoSession()
            .initMocks(this)
            .strictness(Strictness.STRICT_STUBS)
            .startMocking();
...
```
as first call in the setup method and 
```java
    @After
    public void cleanup() throws IOException {
        ...
        mockito.finishMocking();
    }
```
as last call in the teardown method.

With this the unit test becomes:
```java
    @Test
    public void shouldNotCheckpointAfterRestorationWhenExactlyOnceEnabled() {
        final ProcessorStateManager processorStateManager = mock(ProcessorStateManager.class);
        when(processorStateManager.taskType()).thenReturn(TaskType.ACTIVE);
        when(processorStateManager.taskId()).thenReturn(taskId);
        recordCollector = mock(RecordCollectorImpl.class);

        task = createStatefulTask(createConfig(EXACTLY_ONCE_V2, "100"), true, processorStateManager);
        task.initializeIfNeeded();
        task.completeRestoration(noOpResetter -> { });
        verify(processorStateManager, never()).checkpoint();
        verify(processorStateManager, never()).changelogOffsets();
        verify(recordCollector, never()).offsets();
    }
```

If you like you could extract this part into a method and reuse it in both unit tests (that is basically what you originally did, but w\o spy):
```java
        final ProcessorStateManager processorStateManager = mock(ProcessorStateManager.class);
        when(processorStateManager.taskType()).thenReturn(TaskType.ACTIVE);
        when(processorStateManager.taskId()).thenReturn(taskId);
```